Introduce OmitOperation

hnguyen08 opened the pull request.

hnguyen08 committed:
Introduce OmitOperation

hnguyen08 committed:
Introduce OmitOperation

Commit:
Standard update to build files

hnguyen08 committed:
Code review changes:
- Reintroduce old JsonPatch.fromJson API
- RegistryBasedJsonPatchFactory
- Don't default operations upon construction of RegistryBasedJsonPatchFactoryBuilder
- Rename RFC6902* -> Standard*
- Other minor changes

hnguyen08 committed:
Merge branch 'omit' of github.com:hnguyen08/json-patch into omit

hnguyen08 committed:
Made a few changes to help a client easily integrate with the new extended functionality:
- Fixed bug where defaultFactory wasn't initialized
- Added extendedFactory
- Tests for the above

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatch.java:
```diff
@@ -123,11 +123,9 @@ public JsonPatch(final List<JsonPatchOperation> operations)
     public static JsonPatch fromJson(final JsonNode node)
-        throws IOException
-    {
+            throws IOException, JsonPatchException {
```
this smells, usually you have one exception type coming out and either wrap the IOException or don't return a more specifically typed one.

huggsboson replied:
Also JsonProcessingException?

hnguyen08 replied:
Sounds good.  JsonProcessingException extends IOException, so that will be handled as well.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchFactoryUtil.java:
```diff
@@ -0,0 +1,55 @@
+public class JsonPatchFactoryUtil {
+    private static Map<String, Class<? extends JsonPatchOperation>> DEFAULT_OPERATIONS =
+        Collections.unmodifiableMap(
+            new HashMap<String, Class<? extends JsonPatchOperation>>() {{
```
http://stackoverflow.com/questions/924285/efficiency-of-java-double-brace-initialization

huggsboson replied:
ignore it, it isn't an issue I was just curious because I hadn't looked into double brace init.

hnguyen08 replied:
Thanks for looking into it.  I think it is much more natural for the developer with the double brace initialization, and the performance impact isn't too large.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchFactoryUtil.java:
```diff
@@ -0,0 +1,55 @@
+ * has all the RFC6902 operations registered.
+ */
+public class JsonPatchFactoryUtil {
+    private static Map<String, Class<? extends JsonPatchOperation>> DEFAULT_OPERATIONS =
```
final?

hnguyen08 replied:
No, you can.  Done.  In fact the class can be final.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchFactoryUtil.java:
```diff
@@ -0,0 +1,55 @@
+                put("test", TestOperation.class);
+            }}
+        );
+    private static Map<String, Class<? extends JsonPatchOperation>> EXTENDED_OPERATIONS =
```
final?

hnguyen08 replied:
Done

huggsboson commented on src/test/java/com/github/fge/jsonpatch/AddOperationTest.java:
```diff
@@ -22,11 +22,11 @@
         throws IOException
     {
-        super("add");
+        super("add", AddOperation.class);
```
I really do think that having an
interface JsonPatchOperation {
  public String getOpKey();
  public Class<? extends JsonPatchOperationImpl> getImpl();
}

Or something would improve accidentally mis-matching impls with op keys, but I'm willing to let it go.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchFactoryUtil.java:
```diff
@@ -0,0 +1,34 @@
+                put("test", TestOperation.class);
+            }}
+        );
+    private static JsonPatchFactory DEFAULT_FACTORY;
```
isn't this null?

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchFactoryUtil.java:
```diff
@@ -0,0 +1,34 @@
+ * has all the RFC6902 operations registered.
+ */
+public class JsonPatchFactoryUtil {
+    private static Map<String, Class<? extends JsonPatchOperation>> DEFAULT_OPERATIONS =
```
I assume you can't make this final?

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+import com.google.common.collect.Iterables;
+
+/**
+ *
```
Doc

hnguyen08 replied:
Done

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+    public OmitOperation(@JsonProperty("path") final JsonPointer path,
+                         @JsonProperty("value") final JsonNode value)
+    {
+        super("omit", path, value);
```
this makes me think we have the "OpKey" hanging off the PathValueOperation (and ideally all the way up to JsonPatchOperation) which means we should be able to use the "omit" here for the key in the hashMap instead of double specifying it.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+            if (EQUIVALENCE.equivalent(node, value)) {
+                return MissingNode.getInstance();
+            } else {
+                return node.deepCopy();
```
why deep copy here?  I feel like deep copying when we're about to modify is the right move.

hnguyen08 replied:
Good point, deep copy at the beginning makes it harder to make a mistake.  This was fge's style and I didn't take a step back to look at it from scratch.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+    {
+        if (path.isEmpty()) {
+            if (EQUIVALENCE.equivalent(node, value)) {
+                return MissingNode.getInstance();
```
yowza... So you can just omit the entire json value? Is this what remove does with empty path?

hnguyen08 replied:
Yeah, empty path means root of the target JSON node.  So, in "remove" op it means remove the entire JSON node.  A similar idea can be applied here.  Didn't think of it that way, but it is indeed scary, although defined behavior.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+                return node.deepCopy();
+            }
+        }
+        if (path.path(node).isMissingNode())
```
you basically have path.path(node) twice (the below one is path.path(ret)) maybe re-use the result?

hnguyen08 replied:
Done

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+        if (EQUIVALENCE.equivalent(path.path(ret), value)) {
+            final JsonNode parent = path.parent().get(ret);
+            final String rawToken = Iterables.getLast(path).getToken().getRaw();
+            if (parent.isObject())
```
So does the fact that we cover the "" path case and missing path case make it so we basically have to be dealing with either an object or array?

hnguyen08 replied:
Yes, I believe so.  If the target is a primitive like string, number, boolean, null, then any JSON pointer other than "" will result in MissingNode.  It follows the pattern of RemoveOperation.

It is not explicit in the spec, but this line defines this behavior:
Implementations will evaluate each reference token against the
   document's contents and will raise an error condition if it fails to
   resolve a concrete value for any of the JSON pointer's reference
   tokens. 

Since in the spec, evaluating a reference token is defined by only the array and object cases, then evaluating a reference token against any other type will result in an error.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/OmitOperation.java:
```diff
@@ -0,0 +1,54 @@
+        final JsonNode ret = node.deepCopy();
+        if (EQUIVALENCE.equivalent(path.path(ret), value)) {
+            final JsonNode parent = path.parent().get(ret);
+            final String rawToken = Iterables.getLast(path).getToken().getRaw();
```
how does the getRaw() interact with ~1 and ~0?

hnguyen08 replied:
It should decode ~1 and ~0 from the JSON pointer into a "raw" string, i.e. / and ~ respectively.  There are a few tests for ReferenceToken here: https://github.com/fge/jackson-coreutils/blob/master/src/test/java/com/github/fge/jackson/jsonpointer/ReferenceTokenTest.java

huggsboson commented on src/main/java/com/github/fge/jsonpatch/RegistryBasedJsonPatchFactory.java:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Builds a JsonPatchFactory based on the patch operations we want to register
+     */
+    public static class RegistryBasedJsonPatchFactoryBuilder {
```
why not just put this at the top level and remove the nesting/static?

huggsboson replied:
or at least call it only Builder so you get some benefit from the nesting.

hnguyen08 replied:
I like the Builder name.  I put it inside the class because I wanted a private constructor for the actual Factory, and force construction through the Builder.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/RegistryBasedJsonPatchFactory.java:
```diff
@@ -0,0 +1,89 @@
+            this.operations = new HashMap<String, Class<? extends JsonPatchOperation>>();
+        }
+
+        public RegistryBasedJsonPatchFactoryBuilder addOperation(final String operationName, final Class<? extends JsonPatchOperation> c) {
```
I would add getOp() to JsonPatchOperation base class, and move to taking just a final Class<? extends JsonPatchOperation> operation, and then call operations.put(operation.getOp(), operation)

huggsboson commented on src/main/java/com/github/fge/jsonpatch/RegistryBasedJsonPatchFactory.java:
```diff
@@ -0,0 +1,89 @@
+        }
+        String op = node.get("op").asText();
+        if (this.operations.containsKey(op)) {
+            JsonPatchOperation parsedOp = JacksonUtils.getReader().withType(this.operations.get(op)).readValue(node);
```
Did you look into maybe using a custom deserializer instead?
http://www.dilipkumarg.com/dynamic-polymorphic-type-handling-jackson/

hnguyen08 replied:
The custom type deserializer makes sense.  We cannot use #2 with the @JsonTypeResolver annotation because it uses annotations and is not configurable by the developer.  So it seems it would work if we use a combination of: writing a custom TypeDeserializer, and registering the TypeResolver to use this TypeDeserializer (info here: http://stackoverflow.com/questions/12353774/how-to-customize-jackson-type-information-mechanism).

I can play around with it after making the smaller changes from this review.

huggsboson commented on src/test/java/com/github/fge/jsonpatch/JsonPatchFactoryTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    public JsonPatchFactoryTest() {
+        factory = new RegistryBasedJsonPatchFactory.RegistryBasedJsonPatchFactoryBuilder()
+                .addOperation("add", AddOperation.class)
```
mock the operation?

huggsboson commented on src/test/resources/jsonpatch/extended/omit.json:
```diff
@@ -0,0 +1,41 @@
+{
+    "errors": [
```
Some test cases on:
1. path "" with matching value?
1. path "" with simple values 7 and "abc"
2. path / with simple values

hnguyen08 replied:
Done

hnguyen08 committed:
More code review changes:
- Created *OperationFactory classes
- Minor style changes
- made some things final

huggsboson commented on src/main/java/com/github/fge/jsonpatch/JsonPatchOperationFactoryBase.java:
```diff
@@ -0,0 +1,28 @@
+ * The JsonPatchOperationFactoryBase implements a few common operations
+ * for all JsonPatchOperationFactorys.
+ */
+public abstract class JsonPatchOperationFactoryBase implements JsonPatchOperationFactory
```
I'd be tempted to do this as a Helper instead of a Base.

huggsboson commented on src/main/java/com/github/fge/jsonpatch/diff/DiffOperation.java:
```diff
@@ -137,7 +137,7 @@ JsonPatchOperation toOperation(final DiffOperation op)
             JsonPatchOperation toOperation(final DiffOperation op)
             {
-                return new MoveOperation(op.from, op.path);
+                return new MoveOperation("move", op.from, op.path);
```
Is this a different MoveOperation?  When did it pick up an arg contructor?

hnguyen08 replied:
I was playing around and forgot to remove (and run tests).

huggsboson commented:
Looks mostly good to me.  Anyway you can fix the JsonPatchOperation interface thing and then get this patchset rebased on top of head (to eliminate my grails changes) so I can do a quick scan through?

hnguyen08 committed:
Converted JsonPatchOperation to interface, and got rid of a duplicate specification of operation names

hnguyen08 commented:
Pushed final review changes.  Going to rebase and squash (maybe push it to a different PR rather than force push so we keep this history)

hnguyen08 commented:
Merged change up here: #2
